### PR TITLE
Add support for new/delete with alignment that came along in c++17 wi…

### PR DESCRIPTION
**Pull request by baldersheim**
…th gcc 9

@vekterli PR

### CODE REVIEW

File: vespamalloc/src/tests/test1/new_test.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    struct S {
+        int a;
+        long b;
+        int c;
+    };
+    static_assert(sizeof(S) == 24, "sizeof(S) == 16");
+    auto s = std::make_unique<S>();
+    cmp(s.get(), &s->a);
+    cmp(s.get(), 8, &s->b);
+    cmp(s.get(), 16, &s->c);
```
Comment (vekterli) [inline vespamalloc/src/tests/test1/new_test.cpp:24]: For these tests, should we additionally check the returned pointer alignment itself? The struct _field_ alignments should be up to the compiler to get right. Might be enough to add a templated checker function that verifies  `(uintptr_t(s.get()) % alignof(S)) == 0`?

**Review by vekterli — APPROVED** [review]
:+1: a couple minor comments for the tests

File: vespamalloc/src/tests/test1/new_test.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    cmp((static_cast<const char *>(base) + offset), p);
+}
+
+TEST("verify new with normal alignment") {
+    struct S {
+        int a;
+        long b;
+        int c;
+    };
+    static_assert(sizeof(S) == 24, "sizeof(S) == 16");
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespamalloc/src/tests/test1/new_test.cpp:20]: Fixed

File: vespamalloc/src/tests/test1/new_test.cpp (hunk 1 above)
Comment (baldersheim) [reply vespamalloc/src/tests/test1/new_test.cpp:24]: Done, and I also remembered that only vespamalloc is naturally aligned. vespamallocd and vespamallocdst is not as they add preamble and postamble. So they need to be propagated.

File: vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h [hunk 3]
```diff
@@ -45,6 +59,12 @@ class MemBlockBoundsCheckBaseTBase : public CommonT<5>
     void verifyFill() const __attribute__((noinline));
 
     void setSize(size_t sz) { static_cast<uint64_t *>(_ptr)[0] = sz; }
+    static size_t preambleOverhead(std::align_val_t alignment) {
+        return std::max(4*sizeof(unsigned), size_t(alignment));
+    }
+    static size_t preambleOverhead() {
+        return 4*sizeof(unsigned);
+    }
```
Comment (vekterli) [inline vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h:67]: Both these could be `constexpr` (though the compiler will do this in practice anyway)

File: vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h [hunk 4]
```diff
@@ -121,13 +141,23 @@ class MemBlockBoundsCheckBaseT : public MemBlockBoundsCheckBaseTBase
...
 protected:
+    static size_t postambleOverhead() {
+        return sizeof(unsigned) + StackTraceLen*sizeof(void *);
+    }
+    static size_t overhead() {
+        return preambleOverhead() + postambleOverhead();
+    }
+    static size_t overhead(std::align_val_t alignment) {
+        return preambleOverhead(alignment) + postambleOverhead();
+    }
```
Comment (vekterli) [inline vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h:160]: These could also be `constexpr`

File: vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h [hunk 5]
```diff
@@ -45,6 +59,12 @@ class MemBlockBoundsCheckBaseTBase : public CommonT<5>
     void verifyFill() const __attribute__((noinline));
 
     void setSize(size_t sz) { static_cast<uint64_t *>(_ptr)[0] = sz; }
+    static size_t preambleOverhead(std::align_val_t alignment) {
+        return std::max(4*sizeof(unsigned), size_t(alignment));
```
Comment (vekterli) [inline vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h:63]: Could also be `return std::max(preambleOverhead(), size_t(alignment));`

**Review by vekterli — APPROVED** [review]
:+1:

File: vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h (hunk 3 above)
Comment (baldersheim) [reply vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h:67]: Done

File: vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h (hunk 4 above)
Comment (baldersheim) [reply vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h:160]: Done

File: vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h (hunk 5 above)
Comment (baldersheim) [reply vespamalloc/src/vespamalloc/malloc/memblockboundscheck.h:63]: Done

**Pull request merged**